add support for "cd -"

AtomToast opened the pull request:
Not sure if you are interested in prs but I liked the idea of this project and started building a bit on top of it. Would love a code review!
I would also want to keep working on this more in the future

AtomToast committed:
fix memory leak

AtomToast committed:
flesh out cd more

"cd -" works now and there is now also parsing of flags. They aren't
taken into account for anything yet though.

In order to make this work I also implemented some functionality for
setting environmental variables in "export".

g-w1 commented:
Thanks so much for the contribution! This is my first pull request in a project that I made so I am really excited 😃 . I will give some review.

g-w1 commented on src/main.zig:
```diff
@@ -1,7 +1,11 @@
+const CdFlags = enum { L, P };
+
 const Argv = std.ArrayList([]const u8);
+const Flags = std.ArrayList(CdFlags);
```
Not all flags would be CdFlags. IMO, you can just replace Flags in the code with std.ArrayList(CdFlags) if you want. If not, please rename this to show that it is cd.

AtomToast replied:
This is something I was unsure about since I was also thinking about perhaps moving each builtin into their own file.
But yeah, this probably doesn't deserve a top level declaration. At least here

AtomToast replied:
flags are gone for now

g-w1 commented on src/main.zig:
```diff
@@ -80,18 +115,56 @@ fn handleBuiltin(argv: [][]const u8, ally: *std.mem.Allocator) !bool {
+
+        var end_i: usize = line_i;
+        while (envvar.*[end_i] != 0) : (end_i += 1) {}
+        // This may be a memory leak
```
Please test this to see if it is a leak. Just run the export command and see if it leaks after you exit the shell :).

AtomToast replied:
I am not too sure how I would test this. At least there is no error shown once I exit the shell after running export. Would I need to write a test with the testing allocator?

g-w1 replied:
Ah, this will help :)
```
    var alloc = std.heap.GeneralPurposeAllocator(.{}){};
    var gpa = &alloc.allocator;
```
After this, put
```
defer _ = gpa.deinit();
```
This will show if there are leaks in the application when it exits.

g-w1 replied:
If there is nothing shown when it exits, it means that there are no leaks. Zig is awesome!

AtomToast replied:
Hm, it seems to be the alloc.deinit() method you mean but I can't really seem to get it to work quite yet

g-w1 replied:
If there is no output, it means it worked (you got it!)

g-w1 commented on src/main.zig:
```diff
@@ -80,18 +115,56 @@ fn handleBuiltin(argv: [][]const u8, ally: *std.mem.Allocator) !bool {
+                try stdout.print("{s}\n", .{envvar});
+            }
+        } else {
+            const eql_position = std.mem.indexOf(u8, argv[1], "=");
```
Please use a loop here so that we can do stuff like 
`export A=1 B=2`

AtomToast replied:
true, this is something I meant to do but ended up forgetting

AtomToast replied:
this should work now

g-w1 commented on src/main.zig:
```diff
@@ -66,11 +70,42 @@ fn handleBuiltin(argv: [][]const u8, ally: *std.mem.Allocator) !bool {
+                            },
+                        }
+                    }
+                    // TODO: make the flags do sth
```
Please do not introduce the flags if you do not make them do something.

AtomToast replied:
I just added them in since it was kind of already part of parsing the leading dash. But to be honest, I don't think I quite understand what the flags do yet with their "logical" and "physical" interpretation of dots

g-w1 replied:
If you don't understand them, then we do not need them :^)

AtomToast replied:
fair, I just thought the end goal was posix compliance. But might as well focus on the stuff I actually use first

g-w1 replied:
Yep, right now, the goal is a working shell and good code. The end goal is posix compliance :)

AtomToast replied:
flags are gone for now

g-w1 commented on src/main.zig:
```diff
@@ -80,18 +115,56 @@ fn handleBuiltin(argv: [][]const u8, ally: *std.mem.Allocator) !bool {
+                const word = argv[1][pos + 1 ..];
+                try envExport(ally, name, word);
+            } else {
+                try shigError("cd: TODO export existing variables", .{});
```
see std.process.getEnvMap. It returns a std.BufMap which allows you to `get` or `set` on it. Set copies the key and value into the bufmap so you can free it after, reducing the chance of a memory leak. :)

g-w1 replied:
Also, we are not in cd, we are in `export` :)

AtomToast replied:
ah, std.process.getEnvMap seems to be exactly what I have been looking for all night but couldn't find. I really wasn't happy with the current solution so that should hopefully help. It does not seem to update os.environ though so I should probably create a global object somehow.

Will fix that typo :+1:

AtomToast replied:
the comment should be resolved

g-w1 replied:
Feel free to resolve it if you have resolved the issue. (This is for any repo).

AtomToast replied:
okay, I am used to the person that opened the comment also closing it

g-w1 reviewed (commented):
Overall, very nice work!

AtomToast committed:
fix typo

AtomToast committed:
parse all arguments in export

AtomToast committed:
pimp up prompt

make the prompt more distinct from the background to make visual parsing
of terminal output easier

g-w1 commented on src/main.zig:
```diff
@@ -45,8 +49,10 @@ pub fn main() anyerror!void {
+    var buf = [_]u8{0} ** 100;
     const stdout = std.io.getStdOut();
-    try stdout.writer().writeAll("(shig)> ");
+    const cwd = std.os.getcwd(&buf);
```
Peplace this with `std.process.getCwdAlloc(` that takes an allocator please. This way we won't have a problem if the path is more than 100 chars long.
Remember to free it.

AtomToast replied:
I should probably do the same in cd()

AtomToast replied:
hm but then I'd end up passing an allocator to cd again which I was actually quite happy about now having to do

g-w1 replied:
It is better to have correct behavior than clean code.

AtomToast replied:
true

AtomToast committed:
replace std.os.environ with std.process.getEnvMap

AtomToast committed:
prompt: use getCwdAlloc

this prevents issues with prompts longer than 100 characters

g-w1 commented on src/main.zig:
```diff
@@ -7,10 +7,15 @@ const CdFlags = enum { L, P };
     var gpa = &alloc.allocator;
+    defer _ = alloc.deinit();
     const stdout = std.io.getStdOut().writer();
+    env_map = try std.process.getEnvMap(gpa);
```
Make sure to deinit this so that we do not get any leaks.

AtomToast replied:
this should be fixed now. I wonder why the gpa doesn't detect this. Is there many some config setting I need to sett/

AtomToast committed:
cd: remove unused flags

cd flags are postponed for now

AtomToast committed:
env_map: add deinitialisation

AtomToast committed:
cd: use getCwdAlloc

this helps prvent issues with long paths

AtomToast commented:
okay, this would be done then from my side. I think splitting up the files is not something fitting in here
